### PR TITLE
[inductor][xpu] Fix convolution_backward accuracy for mixed channels-last / contiguous inputs

### LINKED ISSUE #1877: Torchbench model functorch_dp_cifar10  got fail_accuracy
### 🐛 Describe the bug

Torchbench model `squeezenet1_1`(bfloat16_inference) and `functorch_dp_cifar10` (float16_train/float32_train/bfloat16_train)  got fail_accuracy.

```
C:\pytorch>python benchmarks\dynamo\torchbench.py --accuracy -d xpu -n10 --backend=inductor --cold-start-latency --inference  --bfloat16  --only squeezenet1_1
loading model: 0it [00:01, ?it/s]
xpu  eval  squeezenet1_1
E0723 13:39:24.765000 8228 site-packages\torch\_dynamo\utils.py:2998] RMSE (res-fp64): 0.06339, (ref-fp64): 0.01168 and shape=torch.Size([4, 1000]). res.dtype: torch.bfloat16, multiplier: 3.000000, tol: 0.000100, use_larger_multiplier_for_smaller_tensor: 0
fail_accuracy




C:\pytorch>python benchmarks\dynamo\torchbench.py --accuracy -d xpu -n10 --backend=inductor --cold-start-latency --training  --bfloat16  --only functorch_dp_cifar10
loading model: 0it [00:01, ?it/s]
xpu  train functorch_dp_cifar10
E0723 13:41:50.568000 3036 site-packages\torch\_dynamo\utils.py:2998] RMSE (res-fp64): 0.00109, (ref-fp64): 0.00027 and shape=torch.Size([64]). res.dtype: torch.bfloat16, multiplier: 3.000000, tol: 0.001000, use_larger_multiplier_for_smaller_tensor: 0
E0723 13:42:38.554000 3036 site-packages\torch\_dynamo\utils.py:2859] Accuracy failed for key name bn1.bias.grad
fail_accuracy

```

### Versions

- Both PT2.8 (03eb1e40f9ddf09cb9eef86ace74332e87f11a79, v2.8.0-rc2) and PT2.7(b1940b5867e40e40ebdce4db76f76d3d0b71d3f4, v2.7.0-rc2) had this issue (BMG Windows).

- Driver: 32.0.101.6913

### PR DESCRIPTION
**Pull request by DamJanusz**
Fixes: https://github.com/intel/torch-xpu-ops/issues/1877
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           
Inductor's layout optimization converts conv weights to channels-last in the forward graph, but the input tensor saved for backward retains its original contiguous strides. This creates a mixed-format scenario where `convolution_backward` receives a channels-last weight but contiguous input/grad_output.                                                                                                                                                                                                                                                                                  
                                                                                                                                                                                                                                                                                                      
On XPU, this mixed format produces incorrect gradient results — particularly with 1×1 spatial dimensions, where contiguous strides `(C,1,1,1)` differ from channels-last strides `(C,1,C,C)` despite covering the same memory.                                                                      
                                                                                                                                                                                                                                                                                                      
This PR adds a custom layout constraint for aten.convolution_backward on XPU that:
  - Checks the tensor device at runtime and early-exits for non-XPU tensors
  - Compares actual strides against canonical channels-last strides (via make_channels_last_strides_for) to detect mixed-format inputs - this correctly handles the 1×1 case where is_contiguous() and is_contiguous(channels_last) are both True
  - Normalises only the non-CL inputs to explicit channels-last strides using require_exact_strides (rather than require_channels_last, which treats size-1 dimensions as don't-care)                                                                                                                                
                                                                                                                                                                     
 The constraint function is selected at import time via torch.xpu._is_compiled(); non-XPU builds use the existing constrain_to_fx_strides with zero overhead.

  Test plan
  - test_conv2d_backward_mixed_memory_format_1x1 — 1×1 spatial dims where contiguous/CL strides diverge
  - test_conv2d_backward_mixed_memory_format — general spatial dims (14×14) with mixed CL weight + contiguous input

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01 @gujinghui @PenghuiCheng @jianyuh @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal @mcarilli @ptrblck @leslie-fang-intel @voznesenskym @penguinwu @EikanWang @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98

### CODE REVIEW

**Comment by guangyey** [comment]
It seems the test case failed in cpu

File: test/inductor/test_torchinductor.py [hunk 1]
```diff
@@ -5329,6 +5329,74 @@ def fn(grad_output, inp, weight):
...
+                [1, 1],
+                False,
+                [0, 0],
+                1,
+                [True, True, True],
+            )
+
+        # On XPU the layout constraint normalises all inputs to channels-last
+        # for correctness, which changes the computation order vs eager and
+        # introduces small numerical differences.
```
Comment (guangyey) [inline test/inductor/test_torchinductor.py]: Could you help me understand why reorder to channel last would introduce numerical difference.

File: torch/_inductor/lowering.py [hunk 2]
```diff
@@ -3189,6 +3190,44 @@ def apply_constraint(arg, fx_arg):
...
+    has_channels_last = False
+    for arg, fx_arg in zip(args, fx_node.args):
+        if not _is_tensor_irnode(arg):
+            continue
+        meta_val = fx_arg.meta.get("val")
+        if meta_val is None or meta_val.dim() != 4:
+            continue
+        if (
+            meta_val.is_contiguous(memory_format=torch.channels_last)
+            and not meta_val.is_contiguous()
```
Comment (guangyey) [inline torch/_inductor/lowering.py]: I don't understand here, for 1x1 spatial dims, here `has_channels_last` still be `False`, how do we relayout it.

File: torch/_inductor/lowering.py [hunk 3]
```diff
@@ -3189,6 +3190,44 @@ def apply_constraint(arg, fx_arg):
...
+        ):
+            has_channels_last = True
+            break
+
+    if has_channels_last:
```
Comment (guangyey) [inline torch/_inductor/lowering.py]: The following logic can't apply to 1x1 spatial dims due to `has_channels_last` is `False`, why this PR could fix the issue.

File: torch/_inductor/lowering.py (hunk 2 above)
Comment (DamJanusz) [reply torch/_inductor/lowering.py]: This was a bug in the previous version. With 1×1 spatial dims, both is_contiguous() and is_contiguous(channels_last) return True, so has_channels_last could never become True and the constraint never fired.

File: torch/_inductor/lowering.py (hunk 3 above)
Comment (DamJanusz) [reply torch/_inductor/lowering.py]: Same issue as above, the old detection logic was insufficient for 1×1.

File: test/inductor/test_torchinductor.py (hunk 1 above)
Comment (DamJanusz) [reply test/inductor/test_torchinductor.py]: When the constraint forces all inputs from contiguous to channels-last layout, the backend may select a different convolution algorithm (e.g. a different tiling or reduction order). Floating-point addition isn't associative, so a different accumulation order produces slightly different results - both numerically valid, but not bitwise identical to eager.

**Comment by DamJanusz** [comment]
> It seems the test case failed in cpu

@expectedFailureCodegenDynamic decorator was missing on both tests. The codegen dynamic shapes path can't handle require_exact_strides constraints with symbolic dimensions. This is a pre-existing limitation affecting all fallback ops with exact-stride constraints.

File: torch/_inductor/lowering.py [hunk 4]
```diff
@@ -3187,6 +3188,70 @@ def apply_constraint(arg, fx_arg):
...
+
+    We detect the mixed case by comparing actual strides against canonical
+    channels-last strides, then normalise non-CL 4-D inputs to explicit
+    channels-last strides. Only applied when tensors are on XPU.
+    """
+    args, kwargs = constrain_to_fx_strides(fx_node, *args, **kwargs)
+
+    # Only apply XPU-specific normalisation when tensors are on XPU.
+    is_xpu = False
+    for fx_arg in fx_node.args:
```
Comment (guangyey) [inline torch/_inductor/lowering.py:3482]: `fx_node.args[0]` means `grad_output`, so I think we could use `fx_node.args[0]` directly to detect if XPU tensor.

**Review by guangyey — COMMENTED** [review]
I know the root cause is `1×1 spatial dims where contiguous strides (C,1,1,1) differ from channels-last (C,1,C,C) despite both satisfying is_contiguous() an is_contiguous(channels_last)`
I am just curious why we don't change to `/*exact_match=*/true`
https://github.com/pytorch/pytorch/blob/838676a9d695c331749ef7d5868f3a9224f7e93a/aten/src/ATen/native/ConvUtils.h#L436-L439 for https://github.com/pytorch/pytorch/blob/838676a9d695c331749ef7d5868f3a9224f7e93a/aten/src/ATen/native/mkldnn/xpu/detail/Utils.cpp#L457-L461
It will help handle this semantics on eager `convolution_backward_overrideable` automatically.
https://github.com/pytorch/pytorch/blob/838676a9d695c331749ef7d5868f3a9224f7e93a/aten/src/ATen/native/mkldnn/xpu/Conv.cpp#L533-L538
cc @EikanWang

**Comment by guangyey** [comment]
@DamJanusz Could you please help rebase the code base and resolve the conflicts.
